[Site Isolation] Add support for copying web content in webarchive format

szewai opened the pull request:
#### 4d90d6f12000158908e404e90adc2225a8671a62
<pre>
[Site Isolation] Add support for copying web content in webarchive format
<a href="https://rdar.apple.com/161923006">rdar://161923006</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=300138">https://bugs.webkit.org/show_bug.cgi?id=300138</a>

Reviewed by Ryosuke Niwa.

In existing implementation, when selecting and copying web content, web process creates webarchive for the selected
range (which might include subframe&apos;s content) and send the webarchive data to UI process. When UI process receives the
data it will add it to pasteboard. With site isolation on, web process only has access to local frame content, so
the data web process sends to UI process is missing content for remote subframes.

To fix it, this patch makes web process not only send selected frame webarchive to UI process, but also remote subframe
identifiers and local subframe webarchives. On receiving the message (WebPasteboardProxy::WriteWebArchiveToPasteBoard),
UI process will collect remote subframe archives from other web processes. After collection, UI process combines all
webarchvies to get a full webarchive for the selected content (the same content it will get wihout site isolaiton).

API Test: SiteIsolation.CreateWebArchiveForCopy
              SiteIsolation.CreateWebArchiveNestedFrameForCopy

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::Editor::writeSelectionToPasteboard):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::LegacyWebArchive):
(WebCore::LegacyWebArchive::create):
(WebCore::LegacyWebArchive::createInternal):
(WebCore::LegacyWebArchive::createFromSelection):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.h:
* Source/WebCore/platform/Pasteboard.h:
* Source/WebCore/platform/PasteboardStrategy.h:
* Source/WebCore/platform/mac/PasteboardMac.mm:
(WebCore::Pasteboard::write):
* Source/WebKit/Shared/APIWebArchive.mm:
(API::WebArchive::WebArchive):
* Source/WebKit/Shared/Pasteboard.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/LegacyWebArchiveCallbackAggregator.h: Added.
* Source/WebKit/UIProcess/Cocoa/WebPasteboardProxyCocoa.mm:
(WebKit::WebPasteboardProxy::writeWebContentToPasteboardInternal):
(WebKit::WebPasteboardProxy::writeWebContentToPasteboard):
(WebKit::WebPasteboardProxy::writeWebArchiveToPasteBoard):
(WebKit::WebPasteboardProxy::createOneWebArchiveFromFrames):
* Source/WebKit/UIProcess/WebPasteboardProxy.h:
* Source/WebKit/UIProcess/WebPasteboardProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebPlatformStrategies.cpp:
(WebKit::collectFrameWebArchives):
(WebKit::WebPlatformStrategies::writeWebArchive):
(WebKit::updateContentForWebArchive):
(WebKit::WebPlatformStrategies::writeToPasteboard):
* Source/WebKit/WebProcess/WebCoreSupport/WebPlatformStrategies.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebPlatformStrategies.h:
* Source/WebKitLegacy/mac/WebView/WebArchive.mm:
(-[WebArchive initWithMainResource:subresources:subframeArchives:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, CreateWebArchiveForCopy)):
(TestWebKitAPI::(SiteIsolation, CreateWebArchiveNestedFrameForCopy)):
(TestWebKitAPI::(SiteIsolation, DoAfterNextPresentationUpdate)):

Canonical link: <a href="https://commits.webkit.org/301206@main">https://commits.webkit.org/301206@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0c9f6a0adfe85b806250f24fed29a2d985ccce2c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125219 "16 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44886 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35625 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132070 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77079 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/aee6fcae-4851-41ed-85d5-15694aade1be) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127096 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45577 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53447 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95329 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/63268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2c914316-5d16-4d14-a8dc-8debc65c6cae) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128173 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36386 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111975 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75869 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/08b930bf-0faf-4ea8-bafc-217e156d1b90) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35283 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30143 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75548 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106153 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30368 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134751 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/52028 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39809 "Found 6 new test failures: http/tests/media/modern-media-controls/overflow-support/playback-speed-live-broadcast.html http/tests/media/video-redirect.html http/wpt/fetch/dnt-header-after-redirection.html imported/w3c/web-platform-tests/html/infrastructure/urls/resolving-urls/query-encoding/utf-8.html?include=loading imported/w3c/web-platform-tests/html/infrastructure/urls/resolving-urls/query-encoding/windows-1252.html?include=loading imported/w3c/web-platform-tests/html/semantics/embedded-content/media-elements/loading-the-media-resource/load-events-networkState.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103795 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52463 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108196 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103565 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26379 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48918 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27208 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/49114 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51920 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57699 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51282 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54638 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52976 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/UIProcess/Cocoa/WebPasteboardProxyCocoa.mm:
```diff
@@ -725,6 +753,79 @@ static PasteboardDataLifetime determineDataLifetime(std::optional<WebPageProxyId
+        return completionHandler(nullptr);
+
+    auto callbackAggregator = WebCore::LegacyWebArchiveCallbackAggregator::create(rootFrameIdentifier, WTFMove(localFrameArchives), WTFMove(completionHandler));
+    HashMap<Ref<WebProcessProxy>, Vector<WebCore::FrameIdentifier>> processFrames;
```
I'd call this `processToFrameMap` or `frameByProcess` for clarity.

rniwa commented on Source/WebKit/UIProcess/Cocoa/LegacyWebArchiveCallbackAggregator.h:
```diff
@@ -0,0 +1,79 @@
+
+namespace WebCore {
+
+class LegacyWebArchiveCallbackAggregator final : public ThreadSafeRefCounted<LegacyWebArchiveCallbackAggregator, WTF::DestructionThread::MainRunLoop> {
```
It seems like we can declare this class in Source/WebKit instead? Is there a reason you put this in WebCore instead?

szewai replied:
`LegacyWebArchive` is in WebCore.

rniwa replied:
yeah but it doesn't seem like this aggregator needs to be in WebCore?

szewai replied:
I am not sure where to put the class so I put it side by side to LegacyWebArchive. Do you have a suggestion on where to move it to?

rniwa replied:
Maybe `Source/WebKit/UIProcess/Cocoa`?

rniwa commented on Source/WebCore/loader/archive/cf/LegacyWebArchiveCallbackAggregator.h:
```diff
@@ -0,0 +1,79 @@
+
+    ~LegacyWebArchiveCallbackAggregator()
+    {
+        if (m_callback)
```
Can we ever create an aggregator with nullptr m_callback? I don't think we can???

szewai replied:
Yes, I will change it to assert.

szewai replied:
It seems default `CompletionHandler` does not have a function pointer

rniwa replied:
What do you mean?

szewai replied:
nvm I was meant to say callers might pass in `CompletionHandler { }` in which case `!m_callback` will be true, but given that existing callers should always pass in a valid callback, I think assert is fine.

rniwa commented on Source/WebCore/loader/archive/cf/LegacyWebArchiveCallbackAggregator.h:
```diff
@@ -0,0 +1,79 @@
+
+    WebCore::FrameIdentifier m_rootFrameIdentifier;
+    HashMap<WebCore::FrameIdentifier, Ref<WebCore::LegacyWebArchive>> m_frameArchives;
+    Callback m_callback;
```
Why isn't this a completion handler?

szewai replied:
You mean `CompletionHandler`?
`using Callback = CompletionHandler<void(RefPtr<LegacyWebArchive>&&)>;`

rniwa replied:
oh, that's rather confusing. Can we just call it `CompletionHandlerType` or something?

rniwa commented on Source/WebKit/UIProcess/Cocoa/LegacyWebArchiveCallbackAggregator.h:
```diff
@@ -0,0 +1,79 @@
+        if (!archive)
+            return archive;
+
+        for (auto subframeIdentifier : archive->subframeIdentifiers()) {
```
We should assert that all results are present.

szewai replied:
hmm subframe archive may not be generated successfully, and the existing implementation just skips it and log error (see `RefPtr<LegacyWebArchive> LegacyWebArchive::createInternal(const String& markupString, const ArchiveOptions& options, LocalFrame& frame, Vector<Ref<Node>>&& nodes, NOESCAPE const Function<bool(LocalFrame&)>& frameFilter)`), so it seems we should not assert here.

rniwa replied:
Oh interesting. Perhaps we want to add a release log in that case? That'll help us diagnose bugs in the future.

rniwa commented on Source/WebKit/WebProcess/WebCoreSupport/WebPlatformStrategies.cpp:
```diff
@@ -251,6 +251,49 @@ int64_t WebPlatformStrategies::setStringForType(const String& string, const Stri
+    if (!webFrame)
+        return;
+
+    for (RefPtr<Frame> frame = webFrame->coreFrame(); frame; frame = frame->tree().traverseNext()) {
```
As we discussed on slack, I think we wanna specify `stayWithin` here. Otherwise, we iterate over the subsequent frames multiple times.

rniwa commented on Source/WebKit/WebProcess/WebCoreSupport/WebPlatformStrategies.cpp:
```diff
@@ -251,6 +251,49 @@ int64_t WebPlatformStrategies::setStringForType(const String& string, const Stri
+            LegacyWebArchive::ShouldArchiveSubframes::No
+        };
+        if (RefPtr archive = WebCore::LegacyWebArchive::create(*document, WTFMove(options)))
+            archives.add(localFrame->frameID(), archive.releaseNonNull());
```
Maybe we wanna assert that `add`'s `isNewEntry` is `true` here.

szewai committed:
[Site Isolation] Add support for copying web content in webarchive format
rdar://161923006
https://bugs.webkit.org/show_bug.cgi?id=300138

Reviewed by Ryosuke Niwa.

In existing implementation, when selecting and copying web content, web process creates webarchive for the selected
range (which might include subframe's content) and send the webarchive data to UI process. When UI process receives the
data it will add it to pasteboard. With site isolation on, web process only has access to local frame content, so
the data web process sends to UI process is missing content for remote subframes.

To fix it, this patch makes web process not only send selected frame webarchive to UI process, but also remote subframe
identifiers and local subframe webarchives. On receiving the message (WebPasteboardProxy::WriteWebArchiveToPasteBoard),
UI process will collect remote subframe archives from other web processes. After collection, UI process combines all
webarchvies to get a full webarchive for the selected content (the same content it will get wihout site isolaiton).

API Test: SiteIsolation.CreateWebArchiveForCopy
	      SiteIsolation.CreateWebArchiveNestedFrameForCopy

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/editing/cocoa/EditorCocoa.mm:
(WebCore::Editor::writeSelectionToPasteboard):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
(WebCore::LegacyWebArchive::LegacyWebArchive):
(WebCore::LegacyWebArchive::create):
(WebCore::LegacyWebArchive::createInternal):
(WebCore::LegacyWebArchive::createFromSelection):
* Source/WebCore/loader/archive/cf/LegacyWebArchive.h:
* Source/WebCore/platform/Pasteboard.h:
* Source/WebCore/platform/PasteboardStrategy.h:
* Source/WebCore/platform/mac/PasteboardMac.mm:
(WebCore::Pasteboard::write):
* Source/WebKit/Shared/APIWebArchive.mm:
(API::WebArchive::WebArchive):
* Source/WebKit/Shared/Pasteboard.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/LegacyWebArchiveCallbackAggregator.h: Added.
* Source/WebKit/UIProcess/Cocoa/WebPasteboardProxyCocoa.mm:
(WebKit::WebPasteboardProxy::writeWebContentToPasteboardInternal):
(WebKit::WebPasteboardProxy::writeWebContentToPasteboard):
(WebKit::WebPasteboardProxy::writeWebArchiveToPasteBoard):
(WebKit::WebPasteboardProxy::createOneWebArchiveFromFrames):
* Source/WebKit/UIProcess/WebPasteboardProxy.h:
* Source/WebKit/UIProcess/WebPasteboardProxy.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebPlatformStrategies.cpp:
(WebKit::collectFrameWebArchives):
(WebKit::WebPlatformStrategies::writeWebArchive):
(WebKit::updateContentForWebArchive):
(WebKit::WebPlatformStrategies::writeToPasteboard):
* Source/WebKit/WebProcess/WebCoreSupport/WebPlatformStrategies.h:
* Source/WebKitLegacy/mac/WebCoreSupport/WebPlatformStrategies.h:
* Source/WebKitLegacy/mac/WebView/WebArchive.mm:
(-[WebArchive initWithMainResource:subresources:subframeArchives:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, CreateWebArchiveForCopy)):
(TestWebKitAPI::(SiteIsolation, CreateWebArchiveNestedFrameForCopy)):
(TestWebKitAPI::(SiteIsolation, DoAfterNextPresentationUpdate)):

Canonical link: https://commits.webkit.org/301206@main
